src/net/server/handlers/create_query.rs

```diff
@@ -0,0 +1,264 @@
+
+    async fn from_request(req: &mut RequestParts<B>) -> Result<Self, Self::Rejection> {
+        #[cfg_attr(feature = "enable-serde", derive(serde::Deserialize))]
+        struct FieldStr {
+            field_type: String,
+        }
+
+        let Query(FieldStr { field_type }) = req.extract().await?;
+        let field_size = match field_type.as_str() {
+            Self::FP2_TYPE => Ok(ff::Fp2::SIZE_IN_BYTES),
+            Self::FP31_TYPE => Ok(ff::Fp31::SIZE_IN_BYTES),
+            Self::FP32_BIT_PRIME_TYPE => Ok(ff::Fp32BitPrime::SIZE_IN_BYTES),
+            other => Err(MpcHelperServerError::bad_query_value(&field_type, other)),
+        }?;
+
+        Ok(Self { field_size })
+    }
+}
+
+/// Wraps a [`BodyStream`] and produce a new stream that has chunks of exactly size `size_in_bytes`.
+/// # Errors
+/// If the downstream body is not a multiple of `size_in_bytes`.
+#[pin_project]
+pub struct ByteArrStream {
+    #[pin]
+    body: BodyStream,
+    size_in_bytes: u32,
+    buffered_size: u32,
+    buffered: VecDeque<Bytes>,
+}
+
+impl ByteArrStream {
+    pub fn new(body: BodyStream, size_in_bytes: u32) -> Self {
+        Self {
+            body,
+            size_in_bytes,
+            buffered_size: 0,
+            buffered: VecDeque::new(),
+        }
+    }
+
+    /// returns [`Bytes`] of `size_in_bytes` length from the buffered chunks
+    /// # Panics
+    /// if the total length of byte buffers is smaller than `size_in_bytes`
+    fn take_bytes(
+        size_in_bytes: u32,
+        buffered: &mut VecDeque<Bytes>,
+        buffered_size: &mut u32,
+    ) -> Bytes {
+        assert!(*buffered_size >= size_in_bytes);
+
+        // if the first buffer is large enough, simply split it in-place.
+        // This is O(1) operation, and should be true majority of the time, assuming large size
+        // buffers
+        if buffered[0].len() > size_in_bytes as usize {
+            let out_bytes = buffered[0].split_to(size_in_bytes as usize);
+            *buffered_size -= size_in_bytes;
+            out_bytes
+        } else {
+            // First buffer is too small, so we need to take across buffers.
+            // Will require a memcopy of u8's across buffers in order to create 1 contiguous buffer
+            // to return.
+            // Removes expended buffers after taking
+
+            let mut out_bytes = Vec::with_capacity(size_in_bytes as usize);
+            // this must loop through the bytes buffers because we don't know how many buffers will
+            // be needed to fulfill `size_in_bytes`. e.g. if every buffer had length 1, we'd need to
+            // visit `size_in_bytes` buffers in order to fill `out_bytes`
+            loop {
+                let remaining_bytes = out_bytes.capacity() - out_bytes.len();
+                if remaining_bytes == 0 {
+                    break;
+                }
+                // current `buffer` has more bytes than needed
+                if buffered[0].len() > remaining_bytes {
```

Comment:
this will panic if you only have one element in the buffer with the len < size_in_bytes

Reply:
lets write a test for that as well

Reply:
you loop also does not have a condition to terminate if you don't have enough bytes in the buffer. lets make sure it is fixed as well